## ----setup---------------------------------------------------------------
require(netDx)
require(GenomicRanges)
require(biomaRt) # for fetching gene coordinates

numCores	<- 8L 

outRoot <- sprintf("%s/200129_threeWay",tempdir())
if (file.exists(outRoot)) unlink(outRoot,recursive=TRUE); dir.create(outRoot)

cat("* Setting up sample metadata\n")
phenoFile <- sprintf("%s/extdata/AGP1_CNV.txt",path.package("netDx"))
pheno   <- read.delim(phenoFile,sep="\t",header=T,as.is=T)
colnames(pheno)[1] <- "ID"
head(pheno)

cnv_GR    <- GRanges(pheno$seqnames,IRanges(pheno$start,pheno$end),
                        ID=pheno$ID,LOCUS_NAMES=pheno$Gene_symbols)
pheno <- pheno[!duplicated(pheno$ID),]


## ----setup2--------------------------------------------------------------
# ----------------------
# Feature selection
outDir <- sprintf("%s/feature_select",outRoot)
dir.create(outDir)

cat("* Create pathway range-sets\n")
pathFile <- fetchPathwayDefinitions("February",2018,verbose=TRUE)
pathwayList <- readPathways(pathFile)

# get gene coordinates, use hg18
ensembl <- useMart("ENSEMBL_MART_ENSEMBL",
	dataset="hsapiens_gene_ensembl",
	host="may2009.archive.ensembl.org",
	path="/biomart/martservice",archive=FALSE)
genes <- getBM(attributes=c("chromosome_name",
		"start_position",
		"end_position",
		"hgnc_symbol"),
	mart=ensembl)
genes <- genes[which(genes[,4]!=""),]

gene_GR     <- GRanges(genes[,1],IRanges(genes[,2],genes[,3]),
   name=genes[,4])
path_GRList <- mapNamedRangesToSets(gene_GR,pathwayList)


## ----makepsn-------------------------------------------------------------
cat("* Make CNV-based patient networks\n")
netDir <- sprintf("%s/networks_orig",outDir)
netList <- makePSN_RangeSets(cnv_GR, path_GRList,netDir,verbose=F)

p 	<- countPatientsInNet(netDir,netList, pheno$ID)
tmp	<- updateNets(p,pheno,writeNewNets=FALSE)
p		<- tmp[[1]]
pheno	<- tmp[[2]] 
p_FULL <- p; pheno_FULL <- pheno; rm(p,pheno)
cat("\n* Run clique-filtering three ways\n")


## ----runpred-------------------------------------------------------------
predictClass	<- "case"
buildPredictor_sparseGenetic(p_FULL,pheno_FULL,predictClass,outDir,netDir,
			enrichLabels=TRUE,numPermsEnrich=20L,numCores=8L,featScoreMax=3L)

## ----plotresults---------------------------------------------------------
baseDir_list <- list()
baseDir_list[["cur"]] <- outDir
	plot(0,0,type="n",xlim=c(0,100),ylim=c(0,100),
		las=1, xlab="FPR (%)", ylab="TPR (%)",bty='n',
		cex.axis=1.5)
	
	# pathway scores
	pathList <- list()
	
	for (nm in names(baseDir_list)) {
	baseDir <- baseDir_list[[nm]]

		print(nm)
		inFile <- sprintf("%s/RR_changeNetSum_stats_denEnrichedNets.txt",	
			baseDir)
		dat	<- read.delim(inFile,sep="\t",header=TRUE,as.is=TRUE)
		points(dat$other_pct,dat$pred_pct,
			  col="red",type="o",pch=16,cex=0.5)
	
		tmp <- data.frame(	
			score=dat$score,
			tp=dat$pred_ol,fp=dat$other_ol,
			# "-" that were correctly not called
			tn=dat$other_tot - dat$other_ol,
			# "+" that were not called 
			fn=dat$pred_tot - dat$pred_ol) 
		stats <- netDx::perfCalc(tmp)
		tmp <- stats$stats
		cat(sprintf("%s: PRAUC = %1.2f\n", nm, stats$prauc))
		cat(sprintf("%s: ROCAUC = %1.2f\n", nm, stats$auc))

		# now get pathway score
		ptmp <- read.delim(sprintf("%s/pathway_cumTally.txt",
			baseDir),sep="\t",h=T,as.is=T)
	}