---
title: "De-sparsifying patient mutations by smoothing over gene interaction networks"
author: "Luca Giudice,Shraddha Pai"
package: netDx
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
    %\VignetteIndexEntry{04. De-sparsifying patient mutations by smoothing over gene interaction networks}.
    %\VignetteEngine{knitr::knitr}
    %\VignetteEncoding{UTF-8}
---

# TL;DR
This code block is not evaluated. Need a breakdown? Look at the following sections.
```{r,eval=TRUE}

```

# Introduction

In this example, we classify ovarian serous cysteadenocarcinoma tumours by prognosis (https://doi.org/10.1038/nature10166, GSE82191). We first reduce sparsity by using prior knowledge of a gene interaction network. Conceptually, the logic is that if a patient has a mutation in a given gene, the mutation indirectly impacts interacting genes. The indirect mutation is inferred by label propagation over a gene-gene interaction network. The resulting smoothed network is then used for downstream applications. Hofree et al. showed that indirect mutations inferred by such smoothing improved patient clustering in 4 types of cancer.

This workflow is identical to that of using unmsmoothed somatic mutations (see vignette 3: XXXXX); the only difference is an additional step to smooth binary mutations.


# Setup

```{r,eval=TRUE}
set.seed(8)
suppressWarnings(suppressMessages(require(netDx)))
suppressWarnings(suppressMessages(require(MultiAssayExperiment)))
```

# Data 
 Data were downloaded from Firehose, hosted by the Broad Institute (doi:10.7908/C11G0KM9). Mutations are represented as a binary matrix with rows corresponding to genes and columns to patients; entry [`i`,`j`] is set to one if gene `i` has a somatic mutation (single/di/tri/oligo-nucleotide polymorphism or indel); else it is set to zero.

```{r,eval=TRUE}
genoFile <- paste(path.package("netDx"),"extdata",
                  "OV_mutSmooth_geno.txt",
	sep=getFileSep())
print(genoFile)
geno <- read.delim(genoFile,sep="\t",header=TRUE,as.is=TRUE)

phenoFile <- paste(path.package("netDx"),"extdata",
                   "OV_mutSmooth_pheno.txt",
                   sep=getFileSep())
pheno <- read.delim(phenoFile,sep="\t",header=TRUE,as.is=TRUE)
```

We apply minor modification to avoid R changing hyphens to dots for sample names:
```{r,eval=TRUE}
colnames(geno) <- gsub("-",".",colnames(geno))
pheno$ID <- gsub("-",".",pheno$ID)
rownames(pheno) <- pheno$ID
```

# Smooth mutations over a gene interaction network
Load the gene-gene interaction network, downloaded from the NBS python implementation of Huang JK and Ideker T doi:10.1093/bioinformatics/bty186:
This network is a compact cancer reference network (CRN) that contains only high-confidence interactions specific to cancer. Huang JK et al. showed that the CRN effectively clusters tumor samples of patients distinguishing them by tumor type and time of survival. 
```{r,eval=TRUE}
netFile <- paste(path.package("netDx"),"extdata",
                 "CancerNets.txt",
	sep=getFileSep())
cancerNets <- read.delim(netFile,sep="\t",header=T,as.is=T)
head(cancerNets)
```

Genes not present in the interaction network are excluded.
[LUCA PLS TRY RUNNING WITHOUT THIS CONDITION. WHY THROW OUT GOOD DATA?]


Use `smoothMutations_LabelProp()` to run label propagation over the mutation matrix, thereby smoothing it.
The result of using this strategy on a patient?s binary somatic mutation profile is a non-sparse network-smoothed profile in which the state of each gene is a continuous value which reflects its network proximity to the original mutations.

```{r,eval=TRUE}
message("* Running label prop")
require(doParallel)
# Start the node clusters for parallel propagation
cl <- makeCluster(1L)
registerDoParallel(cl)
prop_net <- smoothMutations_LabelProp(geno,cancerNets,cl,
                                      no_cores=2L)
stopCluster(cl)
```

Finally, binarize the smoothed matrix.
Only the genes which got a high propagation value are set to one, the others are set to zero.
This step ensures that genes which got a low propagation value are not used.
More a gene is distant from the original mutations, more its propagation value is low and more is
considered unreliable and not involved from them.
```{r,eval=TRUE}
#Set the name of the project and future resulting directory
#Apply binarization
genoP <- thresholdSmoothedMutations(
   prop_net,geno,"OV_CancerNets"
   )
```

# Create pathway-level features with binary patient similarity

Smoothed mutations are now grouped at the level of biological pathways. As with other examples, pathways are downloaded from a compilation of curated pathway databases (`.gmt` format). Thereafter, we define pathway-level patient similarity to be binary; i.e. if two patients share a mutation in genes from the same pathway, their mutual similarity is one; else it is zero.

For more details on the individual steps below, see the first vignette related to building a binary predictor.

```{r,eval=TRUE}
#Setup to build the predictor
pathwayList <- readPathways(
   fetchPathwayDefinitions("January",2018)
   )
exprdat <- SummarizedExperiment(genoP, colData=pheno)
objList <- list(genetic=exprdat)
```

Now we define functions for patient similarity:
```{r, eval=TRUE}
makeNets <- function(dataList,groupList,netDir,numCores,...) {
  netList <- c(); netList2 <- c()
  
  # create genetic nets
  if (!is.null(groupList[["genetic"]])) {
	netList <- makeMutNets(dataList[["genetic"]],
		groupList[["genetic"]],
		netDir,numC=numCores)
  }
  cat(sprintf("\t%i genetic-pathway nets\n", length(netList)))
  cat(sprintf("Total of %i nets\n", length(netList)))
  
  return(netList)
}

# g geno matrix, genes by patients (columns) - binary
# pList list of genesets
# outDir - dir where nets are to be written
makeMutNets <- function(g,pList,oDir,numC) {
  g <- t(g) # transpose to have genes as columns
  cl	<- makeCluster(numC)
  registerDoParallel(cl)
  
  numPat <- c()
  netList <- foreach(k=1:length(pList)) %do% {
    idx <- which(colnames(g) %in% pList[[k]])
    
    if (length(idx)>0) {
      has_mut <- rowSums(g[,idx,drop=FALSE])
      has_mutp <- names(has_mut)[which(has_mut>0)]
      
      if (length(has_mutp)>=6) {
        cat(sprintf("%s: %i patients\n", names(pList)[k],
                    length(has_mutp)))
        #numPat <- c(numPat, length(has_mutp))
        pat_pairs <- t(combinat::combn(has_mutp,2));
        pat_pairs <- cbind(pat_pairs,1);
        outFile <- sprintf("%s/%s_cont.txt",oDir,names(pList)[k])
        write.table(pat_pairs, file=outFile,sep="\t",
                    col=FALSE,row=FALSE,quote=FALSE)
        basename(outFile)
      } else NULL
    } else {
      NULL
    }
  }
  stopCluster(cl)
  unlist(netList)
}
```

# Build predictor
Finally, we compile all the data into a `MultiAssayExperiment` object and as before, run the predictor. 

```{r,eval=TRUE}
exprdat <- SummarizedExperiment(genoP, colData=pheno)
objList <- list(genetic=exprdat)

groupList <- list()
groupList[["genetic"]] <- pathwayList #names for groupList and objList now match

dataList <- MultiAssayExperiment(objList,pheno)
```

The predictor call is essentially the same as with other simpler designs:

```{r,eval=TRUE}
outDir <- tempdir()
#Run the predictor as usual
out <- buildPredictor(dataList=dataList,groupList=groupList,
  makeNetFunc=makeNets, ### custom network creation function
  outDir=outDir, ## absolute path
  numCores=1L, featScoreMax=2L, featSelCutoff=1L,
  numSplits=2L, debugMode=TRUE
)
```

# Examine output

This code collects different components of model output to examine the results. 

```{r,eval=TRUE}
numSplits <- 2
st <- unique(colData(dataList)$STATUS)
acc <- c()         # accuracy
predList <- list() # prediction tables

featScores <- list() # feature scores per class
for (cur in unique(st)) featScores[[cur]] <- list()

for (k in 1:numSplits) { 
	pred <- out[[sprintf("Split%i",k)]][["predictions"]];
	# predictions table
	tmp <- pred[,c("ID","STATUS","TT_STATUS","PRED_CLASS",
	                 sprintf("%s_SCORE",st))]
	predList[[k]] <- tmp 
	# accuracy
	acc <- c(acc, sum(tmp$PRED==tmp$STATUS)/nrow(tmp))
	# feature scores
	for (cur in unique(st)) {
	   tmp <- out[[sprintf("Split%i",k)]][["featureScores"]][[cur]]
	   colnames(tmp) <- c("PATHWAY_NAME","SCORE")
	   featScores[[cur]][[sprintf("Split%i",k)]] <- tmp
	}
}
```

Plot the AUROC and AUPR curves:
```{r,eval=TRUE}
predPerf <- plotPerf(predList, predClasses=st)
```

Examine features with the highest scores. Here, these are pathways with somatic mutations that best predict vital status:
```{r,eval=TRUE}
featScores2 <- lapply(featScores, getNetConsensus)
summary(featScores2)

featSelNet <- lapply(featScores2, function(x) {
    callFeatSel(x, fsCutoff=1, fsPctPass=0)
})
print(head(featScores2[["deceased"]]))
```

# Session Info

```{r,eval=TRUE}
sessionInfo()
```
