---
title: "Tumour classification from gene expression and CNV: Building a predictor from scratch"
output: 
  html_document:
    toc: yes
    toc_depth: 4
  html_notebook:
    toc: yes
    toc_depth: 4
  pdf_document: default
  word_document: default
---

This tutorial shows the steps to build a netDx (Ref 1) breast tumour classifier by integrating gene expression and DNA copy number variants (CNV), using a ***custom predictor design***. 

**IMPORTANT!!** If your predictor has a straightforward cross-validation design, you can also create a predictor with a single function call to `buildPredictor()`.  See the netDx tutorial, "Binary survival prediction from clinical and gene expression: One function call".

# TL;DR
All the code in one chunk. This block is not evaluated in the R notebook. 
Want the process broken down? See the following sections.

```{r,eval=FALSE}
# ******
# NOTE: All of the work detailed in this block is automatically done in the function buildPredictor(). 
# New users are strongly encouraged to use that function, and not go through this process.
# ******

# Change this to a local directory where you have write permission
outDir <- sprintf("%s/TCGA_BRCA",getwd())

numCores        <- 2L   # num cores available for parallel processing
JavaMemory        <- 4L   # java memory in Gb
cutoff          <- 1L   # score cutoff for feature-selected networks
nFold           <- 2L   # two-fold CV, so the code will run fast. 
TRAIN_PROP <- 0.67      # fraction of samples to use for training

if (file.exists(outDir)) unlink(outDir,recursive=TRUE)
dir.create(outDir)

# load packages
require(netDx)
require(netDx.examples)
data(TCGA_BRCA)

subtypes<- c("LumA","other")
pheno$STATUS[which(!pheno$STATUS %in% "LumA")] <- "other"

# split samples into train and test
pheno$TT_STATUS <- splitTestTrain(pheno,
    pctT = TRAIN_PROP,setSeed = 42)

# --------------------------------------------------------------------------
# PHASE 1: Feature selection - uses only training samples
# --------------------------------------------------------------------------
# subset phenotype table and data to training samples
pheno_FULL      <- pheno
xpr_FULL        <- xpr
cnv_FULL        <- cnv_GR
pheno           <- subset(pheno,TT_STATUS %in% "TRAIN")
xpr                     <- xpr[,which(colnames(xpr)%in% pheno$ID)]
cnv_GR          <- cnv_GR[which(cnv_GR$ID %in% pheno$ID)]

# define pathways (here we pull from .gmt) for grouping 
pathFile <- sprintf("%s/extdata/Human_160124_AllPathways.gmt",
    path.package("netDx.examples"))
pathwayList <- readPathways(pathFile)

# create directory for networks and for profiles (pathway-level gene expression files)
profDir <- sprintf("%s/profiles",outDir)
netDir <- sprintf("%s/networks",outDir)

# make PSN from gene expression, group by pathways
netList <- makePSN_NamedMatrix(xpr, rownames(xpr),
        pathwayList,profDir,verbose=FALSE,
        numCores=numCores,writeProfiles=TRUE)
netList <- unlist(netList)

# make PSN from CNV, group by pathways (binary similarity)
gene_GR     <- GRanges(genes$chrom,
   IRanges(genes$txStart,genes$txEnd),
   name=genes$name2)
path_GRList <- mapNamedRangesToSets(gene_GR,pathwayList)
names(path_GRList) <- paste("CNV_",names(path_GRList),sep="")
## warning: this step can take 2-5 minutes depending on the
## number of processes running in parallel
netList2 <- makePSN_RangeSets(cnv_GR, path_GRList,profDir,verbose=F)

# Now that all PSN (features) have been created, compile into one database for feature selection
dbDir   <- compileFeatures(profDir, pheno$ID, outDir,numCores=numCores)

## score networks from 0 to "nFold"; repeat once per class
for (g in subtypes) {
    pDir <- sprintf("%s/%s",outDir,g)
        dir.create(pDir)
        pheno_subtype <- pheno

        ## label patients not in the current class as a residual - note that "nonpred" is used
        ## because neither of the classes uses that label. 
        pheno_subtype$STATUS[which(!pheno_subtype$STATUS %in% g)] <- "nonpred"

        resDir    <- sprintf("%s/GM_results",pDir)
        ## query for feature selection comprises of training
        ## samples from the class of interest
        trainPred <- pheno$ID[which(pheno$STATUS %in% g)]
        
        # Function that does the scoring - this takes a while
        runFeatureSelection(trainPred, resDir, dbDir$dbDir,
                nrow(pheno_subtype),verbose=T, numCores=numCores, nFold=nFold,
                JavaMemory=JavaMemory)

        # Compute network tally (from 0 to nFold)
        # file with network weights for each scoring call
        nrank <- dir(path=resDir,pattern="NRANK$")
        pTally          <- compileFeatureScores(paste(resDir,nrank,sep="/"))
        tallyFile       <- sprintf("%s/%s_pathway_CV_score.txt",resDir,g)
        write.table(pTally,file=tallyFile,sep="\t",col=T,row=F,quote=F)
}

# --------------------------------------------------------------------------
# PHASE 2: Classify test samples
# --------------------------------------------------------------------------
pheno <- pheno_FULL
predRes <- list

# Compute similarity of test samples to each class in turn.
for (g in subtypes) {
        pDir <- sprintf("%s/%s",outDir,g)
        # get feature selected net names
        pTally <- read.delim(
                sprintf("%s/GM_results/%s_pathway_CV_score.txt",pDir,g),
                sep="\t",h=T,as.is=T)
        pTally <- pTally[which(pTally[,2]>=cutoff),1]
        pTally <- sub(".profile","",pTally)
        pTally <- sub("_cont","",pTally)

        cat(sprintf("%s: %i pathways\n",g,length(pTally)))
        profDir <- sprintf("%s/profiles",pDir)

        # Make PSN using train & test samples, first for gene xpr, then for CNV
        tmp <- makePSN_NamedMatrix(xpr_FULL,rownames(xpr),
                pathwayList[which(names(pathwayList)%in% pTally)],
                profDir,verbose=F,numCores=numCores,writeProfiles=TRUE)
        tmp <- makePSN_RangeSets(cnv_FULL,
                path_GRList[which(names(path_GRList)%in% pTally)],
                profDir,verbose=FALSE)
        # create db
        dbDir <- compileFeatures(profDir,pheno$ID,pDir,numCores=numCores)

        # query of all training samples for this class
        qSamps<-pheno$ID[which(pheno$STATUS %in% g & pheno$TT_STATUS%in%"TRAIN")]
        qFile<-sprintf("%s/%s_query",pDir,g)
        writeQueryFile(qSamps,"all",nrow(pheno),qFile)

        resFile<-runQuery(dbDir$dbDir,qFile,resDir=pDir)

        predRes[[g]]<-getPatientRankings(sprintf("%s.PRANK",resFile),pheno,g)
}

# Finally, classify each sample based on which class it is more similar to
predClass <- predictPatientLabels(predRes)

# Compute performance metrics
both <- merge(x=pheno,y=predClass,by="ID")
print(table(both[,c("STATUS","PRED_CLASS")]))
pos <- (both$STATUS %in% "LumA")
tp <- sum(both$PRED_CLASS[pos]=="LumA")
fp <- sum(both$PRED_CLASS[!pos]=="LumA")
tn <- sum(both$PRED_CLASS[!pos]=="other")
fn <- sum(both$PRED_CLASS[pos]=="other")
cat(sprintf("Accuracy = %i of %i (%i %%)\n",tp+tn,nrow(both),
                        round(((tp+tn)/nrow(both))*100)))
cat(sprintf("PPV = %i %%\n", round((tp/(tp+fp))*100)))
cat(sprintf("Recall = %i %%\n", round((tp/(tp+fn))*100)))

```

# Introduction
Based on the expression of 50 genes, breast tumours are traditionally classified into one of four broad categories: Luminal A, Luminal B, HER2-enriched, and basal-like (Ref 2). Each category of tumour has different  prognostic value and response to chemotherapy and/or hormone therapy.

Here we build a binary classifier that discriminates between Luminal A and other subtypes (i.e. non-Luminal A). As input data, we have a matrix of processed gene expression and CNV calls. Through this exercise, we will use the following capabilities of netDx:

* Create pathway-level features (patient similarity networks) from gene expression data
* Create pathway-level features from sparse CNV data
* Perform feature selection on the training set, so features have scores
* Assess performance on the hold-out (test) samples

# Step 0. Set up environment
A multi-core compute node will considerably speed the process and is highly recommended. Change the `numCores` parameter to the number of cores on your machine.

``` {r, eval=TRUE} 
# Change this to a local directory where you have write permission
outDir <- sprintf("%s/TCGA_BRCA",getwd())
cat(sprintf("All intermediate files are stored in:\n%s\n",outDir))

numCores        <- 2L   # num cores available for parallel processing
JavaMemory        <- 4L   # java memory in Gb
cutoff          <- 1L   # score cutoff for feature-selected networks
nFold           <- 2L   # two-fold CV, so the code will run fast. 
TRAIN_PROP <- 0.67      # fraction of samples to use for training

if (file.exists(outDir)) unlink(outDir,recursive=TRUE)
dir.create(outDir)
```

Load the netDx software and data packages. Finally, load the breast cancer dataset.

``` {r, eval=TRUE}
require(netDx)
require(netDx.examples)
data(TCGA_BRCA)
```

# Step 1. Split data into training and test sets

```{r, eval=TRUE}

subtypes<- c("LumA")
pheno$STATUS[which(!pheno$STATUS %in% subtypes)] <- "other"
subtypes <- c(subtypes,"other") 

pheno$TT_STATUS <- splitTestTrain(pheno,
    pctT = TRAIN_PROP,setSeed = 42)
```


# Step 2. Feature selection (training samples only)

## 2a. Limit samples/data to training samples

Subset phenotype table and individual data layers to keep only training samples
``` {r, eval=TRUE}
pheno_FULL      <- pheno
xpr_FULL        <- xpr
cnv_FULL        <- cnv_GR
pheno           <- subset(pheno,TT_STATUS %in% "TRAIN")
xpr                     <- xpr[,which(colnames(xpr)%in% pheno$ID)]
cnv_GR          <- cnv_GR[which(cnv_GR$ID %in% pheno$ID)]
```

## 2b. Define groupings (pathways or gene-sets)
Here we load  pathway definitions precompiled into a [.gmt file](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29) from curated databases such as Panther, Reactome, NCI, IOB and KEGG. 

**Updated genesets are available for public download from [the Baderlab Geneset page](http://download.baderlab.org/EM_Genesets/).**

For example, human genome pathway definitions for April 2019 are available under `April_01_209/Human/symbol`, **[as in here](http://download.baderlab.org/EM_Genesets/April_01_2019/Human/symbol/)**.

``` {r, eval=TRUE}
pathFile <- sprintf("%s/extdata/Human_160124_AllPathways.gmt",
    path.package("netDx.examples"))
pathwayList <- readPathways(pathFile)
head(pathwayList)
```

## 2c. Create features (Patient Similarity Networks or PSN)

netDx converts input data into patient similarity networks. For each type of input data (e.g. clinical, gene expression, CNV), use built-in commands to generate PSN. Finally compile them all into a single "database". The database will be used to score features (feature selection). This section shows you how to create PSN from dense gene expression data or sparse CNV data (i.e. genomic events). 

####  Make PSN from gene expression data
From gene expression data, we create one network per cellular pathway. Similarity between two patients is defined as the Pearson correlation of the expression vector; each network is limited to genes for the corresponding pathway.
\par
The function that generates the networks from submatrices of the gene expression data is `makePSN_NamedMatrix()`. In this case, we are generating "profiles", or simply writing submatrices corresponding to the pathways (note the `writeProfiles=TRUE` argument). As these profiles will create completely connected networks with N-choose-2 edges, weaker edges will first be pruned for computational feasibility. Note that `netList` contains the names of networks, rather than the contents; the profiles are written to `profDir`. Profile file names end with `.profile`.

``` {r, eval=TRUE}
profDir <- sprintf("%s/profiles",outDir)
netDir <- sprintf("%s/networks",outDir)

netList <- makePSN_NamedMatrix(xpr, rownames(xpr),
        pathwayList,profDir,verbose=FALSE,
        numCores=numCores,writeProfiles=TRUE)
netList <- unlist(netList)
head(netList)
```

#### Make PSN from copy number variants (sparse)
Similarly, we construct networks based on shared overlap of CNVs. For each cellular pathway, we create a network consisting of patients with CNVs in the member genes of that pathway (or gene-set). The edge weight here is binary; all patients in the network have an edge weight of one. Those not in the network implicitly have a weight of zero.

Genomic events (here, CNVs) need to first be mapped to genes, before being grouped into pathways. This mapping is achieved by `mapNamedRangesToSets()`. The function used to construct networks from genomic events is `makePSN_RangeSets()`. As with the gene-expression nets, CNV nets are written to `profDir`. All input networks must be in the same directory. Interaction network names end with `_cont.txt`.

``` {r, eval=TRUE}
data(genes)
gene_GR     <- GRanges(genes$chrom,
   IRanges(genes$txStart,genes$txEnd),
   name=genes$name2)
path_GRList <- mapNamedRangesToSets(gene_GR,pathwayList)
names(path_GRList) <- paste("CNV_",names(path_GRList),sep="")
## warning: this step can take 2-5 minutes depending on the
## number of processes running in parallel
netList2 <- makePSN_RangeSets(cnv_GR, path_GRList,profDir,verbose=F)
cat(sprintf("CNV: Got %i networks\n",length(netList2)))
```

Let's take a look at CNV-based networks:
``` {r, eval=TRUE}
head(unlist(netList2))
```

## 2d. Compile all features into a single database
Once all our patient networks are constructed, these are integrated into a single database for feature selection.

``` {r, eval=TRUE}
dbDir   <- compileFeatures(profDir, pheno$ID, outDir,numCores=numCores)
```

## 2e. Score networks (or feature selection)
Figure 1B shows the schematic for feature selection. The goal of this step is to extract the networks that are most predictive of a given class. For each subtype, here "LumA" and "other", feature selection is performed once (the large outer `for` loop).

The key functions are:

* `runFeatureSelection()`, which scores the networks
* `compileFeatureScores()`, which loops over all network rank files (or NRANK files) and computes the network score

``` {r, eval=TRUE}
## repeat process for each class
for (g in subtypes) {
    pDir <- sprintf("%s/%s",outDir,g)
    if (file.exists(pDir)) unlink(pDir,recursive=TRUE)
        dir.create(pDir)

        cat(sprintf("\n******\nSubtype %s\n",g))
        pheno_subtype <- pheno

        ## label patients not in the current class as a residual
        pheno_subtype$STATUS[which(!pheno_subtype$STATUS %in% g)] <- "nonpred"
        ## sanity check
        print(table(pheno_subtype$STATUS,useNA="always"))

        resDir    <- sprintf("%s/GM_results",pDir)
        ## query for feature selection comprises of training
        ## samples from the class of interest
        trainPred <- pheno$ID[which(pheno$STATUS %in% g)]

        # Cross validation
        runFeatureSelection(trainPred, resDir, dbDir$dbDir,
                nrow(pheno_subtype),verbose=T, numCores=numCores, nFold=nFold,
                JavaMemory=JavaMemory)

        # patient similarity ranks
        prank <- dir(path=resDir,pattern="PRANK$")
        # network ranks
        nrank <- dir(path=resDir,pattern="NRANK$")
        cat(sprintf("Got %i prank files\n",length(prank)))

    # Compute network score
        pTally          <- compileFeatureScores(paste(resDir,nrank,sep="/"))
        head(pTally)
        # write to file
        tallyFile       <- sprintf("%s/%s_pathway_CV_score.txt",resDir,g)
        write.table(pTally,file=tallyFile,sep="\t",col=T,row=F,quote=F)
}
```

# 3. Classify test samples from selected features
Following feature selection (previous section), we have identified the networks that are predictive of our two classes of interest: LumA and other. For each of these classes, we now create a single database comprising ***only of the feature selected nets*** ; this is equivalent to our trained model for each class. 

We rank the similarity of a test patient to each class. 
For example:

* Rank_LumA: Similarity rank to  training  `LumA` samples
* Rank_other: Similarity to training `other` samples
* Final rank = `max(Rank_LumA,Rank_other)`

The following code block does all these steps:

* `makePSN_NamedMatrix`, `makePSN_RangeSets`: Create patient nets for the feature-selected networks using both training and test samples
* `compileFeatures`: Create the new database from the resulting nets
* `runQuery`: Query to get patient similarity, run once per class
* `getPatientRankings`: Get patient rankings

```{r, eval=TRUE}
pheno <- pheno_FULL
predRes <- list()
for (g in subtypes) {
        pDir <- sprintf("%s/%s",outDir,g)
        # get feature selected net names
        pTally <- read.delim(
                sprintf("%s/GM_results/%s_pathway_CV_score.txt",pDir,g),
                sep="\t",h=T,as.is=T)
        pTally <- pTally[which(pTally[,2]>=cutoff),1]
        pTally <- sub(".profile","",pTally)
        pTally <- sub("_cont","",pTally)

        cat(sprintf("%s: %i pathways\n",g,length(pTally)))
        profDir <- sprintf("%s/profiles",pDir)

        # prepare nets for new db
        tmp <- makePSN_NamedMatrix(xpr_FULL,rownames(xpr),
                pathwayList[which(names(pathwayList)%in% pTally)],
                profDir,verbose=F,numCores=numCores,writeProfiles=TRUE)
        tmp <- makePSN_RangeSets(cnv_FULL,
                path_GRList[which(names(path_GRList)%in% pTally)],
                profDir,verbose=FALSE)
        # create db
        dbDir <- compileFeatures(profDir,pheno$ID,pDir,numCores=numCores)

        # query of all training samples for this class
        qSamps<-pheno$ID[which(pheno$STATUS %in% g & pheno$TT_STATUS%in%"TRAIN")]
        qFile<-sprintf("%s/%s_query",pDir,g)
        writeQueryFile(qSamps,"all",nrow(pheno),qFile)

        resFile<-runQuery(dbDir$dbDir,qFile,resDir=pDir)

        predRes[[g]]<-getPatientRankings(sprintf("%s.PRANK",resFile),pheno,g)
}
```

In the last section, we obtained two similarity ranks for each test patient. Here we use `predictPatientLabels()` to label patients by max rank.

``` {r, eval=TRUE}
predClass <- predictPatientLabels(predRes)
cat("Predicted classes\n")
```

# Step 4. Evaluate performance

``` {r, eval=TRUE}
both <- merge(x=pheno,y=predClass,by="ID")
print(table(both[,c("STATUS","PRED_CLASS")]))
pos <- (both$STATUS %in% "LumA")
tp <- sum(both$PRED_CLASS[pos]=="LumA")
fp <- sum(both$PRED_CLASS[!pos]=="LumA")
tn <- sum(both$PRED_CLASS[!pos]=="other")
fn <- sum(both$PRED_CLASS[pos]=="other")
cat(sprintf("Accuracy = %i of %i (%i %%)\n",tp+tn,nrow(both),
                        round(((tp+tn)/nrow(both))*100)))
cat(sprintf("PPV = %i %%\n", round((tp/(tp+fp))*100)))
cat(sprintf("Recall = %i %%\n", round((tp/(tp+fn))*100)))
```

# sessionInfo

```
sessionInfo()
```

# References
1. Pai et al. (2019). [netDx: A patient classifier based on integration of patient similarity networks.](http://msb.embopress.org/content/15/3/e8497) *Mol Sys Biol* **15**, e8497. 
2. Parker JS et al. (2009) *J Clin Oncol.* **27** (8):1160-7.
