\documentclass{article}

\begin{document}

\title{netDx use case: 4-way classification: Medulloblastoma tumour subtype}
\author{Shraddha Pai}
\date{Last updated: 3 June, 2016}

\maketitle

\section{Introduction}
Northcott et al. (2011) identified four subtypes of primary medulloblastoma tumours based on gene expression profiles, each predictable by levels of ~5 genes (Ref 1). For each tumour, we have gene-level expression data (GSE21140) from the Affymetrix Human Exon 1.0 ST array (gene-level). In this application, our goal is to classify a new tumour into one of the 4 subtypes.
\par
The netDx workflow is shown in Figure 1. We use gene signatures identified by previous research and omit the feature selection step. Instead, GeneMANIA (GM) is used to predict tumour labels. For each of the 4 subtypes, we construct a GM database that contains genes part of the corresponding gene signature. Samples from each subtype are split into training (70\%) and test (30\%). One query is run for each of the four subtypes, using the training samples for the corresponding subtypes as a query. In this way, each test sample is ranked by all 4 databases. After test samples have been ranked by all four GeneMANIA queries , each set of ranks is normalized to range between 0 and 1.  The patient is assigned to the class with the highest rank.

\includegraphics[width=1.0\textwidth]{mblastoma.png}

\section{Set up}
\subsection{Set up working environment}
<<>>=
# change this to a directory to which you have write access
outDir <-  "~/tmp/MB"
dir.create(outDir)

numCores    <- 2L	# number of cores for parallel processing
pctTrain    <- 0.7	# fraction of samples to use for feature selection

require(netDx)
require(netDx.examples)
@

Load the example data
<<>>=
# Load the Medulloblastoma dataset
data(MBlastoma)
@

\subsection{Define gene signatures}

<<>>=
# subtypes and genes predictive of these. From Table 1 of PMC3306784
# http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3306784/table/Tab1/
groupSig <- list(
    WNT=c("WIF1","TNC","GAD","DKK2","EMX2"),
    SHH=c("PDLIM3","EYA1","HHIP","ATOH1","SFRP1"),
    Group3=c("IMPG2","GABRA5","EGFL11","NRL","MAB21L2","NPR3"),
    Group4=c("KCNA1","EOMES","KHDRBS2","RBM24","UNC5D","OAS1")
)
@

\subsection{Split samples into train and test}

For this example we manually separate samples from each of the 4 subtypes into 70\% training and 30\% test
<<>>=
# ---------------------------------------------
# split dataset into train/test in group-wise manner. Note that we 
# include patients not in any particular group because they serve
# as negatives
TT_STATUS <- character(nrow(MB.pheno))
for (g in unique(MB.pheno$STATUS)) {
    idx <- which(MB.pheno$STATUS %in% g)
    status <- rep("TEST",length(idx))
    status[1:(floor(pctTrain * length(idx)))] <- "TRAIN"
    TT_STATUS[idx] <- sample(status, replace=FALSE) # scramble
}
MB.pheno <- cbind(MB.pheno, TT_STATUS=TT_STATUS)
@

\section{Build predictor for each subtype}
In this example, we skip feature selection and move straight to predicting subtypes for test tumours, based on similarity to training samples.

<<>>=
MB.pheno_train <- subset(MB.pheno, TT_STATUS %in% "TRAIN")
@

For each subtype, we build a GeneMANIA database consisting of features for that subtype.  Each database contains all 103 samples. Features are at the levels of genes. The similarity measure is a custom-defined one, here seen in the geneSim function.

<<>>=
# custom similarity measure
geneSim <- function(x) {
    if (nrow(x)>=1) x <- x[1,]
    nm <- colnames(x)
    x <- as.numeric(x)
    n <- length(x)
    rngX  <- max(x)-min(x)
    
    out <- matrix(NA,nrow=n,ncol=n);
    # weight between i and j is
    # wt(i,j) = 1 - (abs(g[i]-g[j])/(max(g)-min(g)))
    # where g is the eMB.xpression vector for each gene
    for (j in 1:n) out[,j] <- 1-(abs((x-x[j])/rngX))
    rownames(out) <- nm; colnames(out)<- nm
    out
}
@

<<>>=
# directories with group-specific predictors
predRes <- list()
## for each subtype
for (g in names(groupSig)){
    pDir <- sprintf("%s/%s",outDir,g)
    if (file.exists(pDir)) unlink(pDir)
    dir.create(pDir)
    
    # each gene has its own PSN
    sigNets <- list()
    for (g2 in groupSig[[g]]) sigNets[[g2]] <- g2
    
    # create patient networks using train & test samples
    # networks are limited to signature genes for this subtype
    idx     <- which(MB.xpr_names %in% groupSig[[g]])
    cat(sprintf("Subtype : %s { %s } => %i measures\n",
                g, paste(groupSig[[g]],collapse=","), length(idx)))
    netDir  <- sprintf("%s/networks",pDir)
    if (!file.exists(netDir)) unlink(netDir)
    
    # this function call creates the PSN
    netList <- makePSN_NamedMatrix(MB.xpr[idx,], MB.xpr_names[idx], 
                                   sigNets,netDir,
                                   simMetric="custom",customFunc=geneSim,
                                   verbose=TRUE)
    
    # create a GeneMANIA database out of these networks
    dbDir <- GM_createDB(netDir, MB.pheno$ID, pDir)
    
    # run a query using training samples for this subtype.
    # get ranking for all patients in the database by running GeneMANIA
    trainSamps <- MB.pheno$ID[which(MB.pheno$TT_STATUS %in% "TRAIN" 
		& MB.pheno$STATUS %in% g)]
    qFile      <- sprintf("%s/query.txt", pDir)
    GM_writeQueryFile(trainSamps, "all", nrow(MB.pheno),qFile)
    resFile <- runGeneMANIA(dbDir$dbDir, qFile, pDir)
    
    # compute ROC curve for each predictor
    predRes[[g]] <- GM_getQueryROC(sprintf("%s.PRANK",resFile),
            MB.pheno, g)
}
@

\section{Evaluate performance}
Once we have run GeneMANIA on all 4 subclasses, the patient gets assigned to the class for which it has the highest ranking.

Take a look at the ROC curves.
<<>>=
par(mfrow=c(2,2))
tmp <- sapply(names(predRes), function(nm){
    x <- predRes[[nm]]
    plot(x$roc,
         main=sprintf("%s: N=%i (AUC=%1.3f)",
                      nm,length(x$roc@x.values[[1]]),x$auc),
         cex.main=0.8)
    })

save(predRes,file=sprintf("%s/predictions.Rdata",outDir))
@


Finally, predict the class of each test sample
<<>>=
predClass 	<- GM_OneVAll_getClass(predRes)
testSamps 	<- merge(x=MB.pheno,y=predClass,by="ID")
@

Compute class match accuracy
<<>>=
rightClass <- testSamps$STATUS == testSamps$PRED_CLASS
numCor <- sum(rightClass); ln <- nrow(testSamps)
cat(sprintf("Overall classifier accuracy = %i of %i  (%i%%)",
            numCor, ln, round((numCor/ln)*100)))
@

Examine class-specific accuracy:
<<>>=
for (g in names(groupSig)){
    idx <- which(testSamps$STATUS %in% g)
    numCor <- sum(rightClass[idx]); ln <- length(idx)
    cat(sprintf("\t%s = %i of %i  (%i%%)\n",
            g, numCor, ln, round((numCor/ln)*100)))
}
@

\section{sessionInfo}
<<>>=
sessionInfo()
@

\section{References}
1. Northcott PA et al. (2011). J Clin Oncol. 29 (11):1408.
\end{document}