---
title: 'Plot netDx results: One function'
output:
  pdf_document: default
  html_notebook:
    toc: yes
    toc_depth: 2
  html_document:
    toc: yes
    toc_depth: 2
  word_document: default
---

# Introduction
This notebook shows you how to plot all netDx predictor results in one function call. The output generated includes:

* **Summary predictor performance:** Plot of average performance (AUROC/AUPR)
* **Detailed predictor performance:** AUROC/AUPR curves for all data splits
* **Detailed statistics:** Network scores and patient predictions in all splits
* **Themes of selected features:** A network visualization of major themes in predictive variables (an EnrichmentMap)
* A **network visualization of patient similarity** using only feature-selected variables

# Software Requirements
Assumes you have all Cytoscape- and CyRest-related dependencies required to install netDx v1.0.

**Note:If Cytoscape is not running, this example will not work!**

Let us see if the required dependencies can be installed and/or loaded:

```{r}
#httr
tryCatch(expr = { library(httr)}, 
          error = function(e) { install.packages("httr")}, finally = library(httr))

#RJSONIO
tryCatch(expr = { library(RJSONIO)}, 
          error = function(e) { install.packages("RJSONIO")}, finally = library(RJSONIO))

#r2cytoscape
tryCatch(expr = { library(r2cytoscape)}, 
          error = function(e) { devtools::install_github('cytoscape/cytoscape-automation/for-scripters/R/r2cytoscape')}, finally = library(r2cytoscape))

# EasycyRest
tryCatch(expr = { library(EasycyRest); detach(package:EascyRest,unload=TRUE)}, 
          error = function(e) { devtools::install_github('BaderLab/Easycyrest/EasycyRest@0.1')}, finally = {})
```

# Set up

```{r,eval=TRUE}
suppressWarnings(suppressMessages(require(netDx)))
suppressWarnings(suppressMessages(require(netDx.examples)))
```

## Load data for plotting
In this example, we use data from The Cancer Genome Atlas (http://cancergenome.nih.gov/), downloaded from the PanCancer Survival project (https://www.synapse.org/#!Synapse:syn1710282). We use gene expression profiles from renal clear cell carcinoma tumours to predict poor and good survival after Yuan et al. (2014) (Refs 1-2). The data consists of 150 tumours. Here we work only with the gene expression profiles generated.

```{r,eval=TRUE}
phenoFile <- sprintf("%s/extdata/KIRC_pheno.rda",path.package("netDx.examples"))
lnames <- load(phenoFile)
head(pheno)
```

```{r}
pathFile <- sprintf("%s/extdata/Human_160124_AllPathways.gmt",
           path.package("netDx.examples"))
pathwayList <- readPathways(pathFile)
```
Filter for the genes measured in this dataset. For this example we have stored the names of genes measured in this dataset. In practice you would get this information from the corresponding input table.

```{r}
xpr_genes <- sprintf("%s/extdata/EMap_input/genenames.txt",
      path.package("netDx.examples"))
xpr_genes <- read.delim(xpr_genes,h=FALSE,as.is=TRUE)[,1]
head(xpr_genes)
```
```{r}
pathwayList <- lapply(pathwayList, function(x) x[which(x %in% xpr_genes)])
```

# Generate results

```{r}
inDir <- sprintf("%s/extdata/KIRC_output",
	path.package("netDx.examples"))
out <- plotAllResults(pheno, inDir,outDir=sprintf("%s/plots",getwd()),
               fsCutoff=10,fsPctPass=0.7,pathwaySet=pathwayList)
```

The EnrichmentMaps and integrated PSN should now be present in Cytoscape, as well as PNG files in the output directory.


# Examine results

**Feature scores:**

* `featScores/<class>_featScores.txt`: Table of feature scores for all splits of nested CV.
* `featScores/<class>_FeatSel_cutoff<fsCutoff>_pct<fsPctPass.txt`: Features selected for this class

**Integrated patient dissimilarity network:**

* `PSN/outputPDN.png`: Integrated patient dissimilarity network
* `PSN/aggregateNet_filterEdgeWt0.00_MEAN.txt`: Aggregate PSN from combining all feature-selected nets
* `PSN/predictor_prunedNet_top0.20.txt`: Pruned dissimilarity network that is the input to the viz in Cytoscape
* `PSN/pool/*txt`: All interaction nets that were input for the aggregate PSN

**Enrichment Map:**

* `EMap/EnrichmentMap_<class>.png`: View of EnrichmentMap for each class
* `EMap/<class>_<yymmdd>.gmt`: GMT file that serves as input for EnrichmentMap in Cytoscape
* `EMap/<class>_nodeAttrs_<yymmdd>.txt`: Node attribute table with "maxScore" column that is mapped to node fill in Cytoscape.

```{r}
dir(sprintf("%s/plots",getwd()),recursive=TRUE)
```

# sessionInfo

```{r}
sessionInfo()
```

# References
1. Yuan, Y. *et al.* (2014) Assessing the clinical utility of cancer genomic and proteomic data across tumor types. *Nat Biotechnol* **32**, 644-52.
2. The Cancer Genome Atlas Research Network (2013). Comprehensive molecular characterization of clear cell renal cell carcinoma. *Nature* **499**, 43-9.
3. Merico, D., Isserlin, R. & Bader, G.D. (2011). Visualizing gene-set enrichment results using the Cytoscape plug-in enrichment map. *Methods Mol Biol* **781**, 257-77.
4. Kucera, M., Isserlin, R., Arkhangorodsky, A. & Bader, G.D. (2016). AutoAnnotate: A Cytoscape app for summarizing networks with semantic annotations. *F1000Res* **5**, 1717.



